File: crypto-ffi/bindings/swift/WireCoreCrypto/WireCoreCrypto/CoreCrypto.swift
import System
@_exported public import WireCoreCryptoUniffi

/// Defines the protocol for a client.
///
public protocol CoreCryptoProtocol: CoreCryptoFfiProtocol {

    /// Instantiate a history client.
    ///
    /// This client exposes the full interface of ``Self``, but it should only be used to decrypt messages.
    /// Other use is a logic error.
    static func historyClient(_ historySecret: HistorySecret) async throws -> Self

    /// Starts a transaction in Core Crypto. If the closure succeeds without throwing an error,
    /// it will be committed, otherwise, every operation performed with the context will be discarded.
    ///
    /// - Parameter block: the closure to be executed within the transaction context.
    /// A ``CoreCryptoContext-swift.protocol`` is provided on which any operations should be performed.
    ///
    /// - Returns: Result value returned from the closure if any.
    ///
    func transaction<Result>(
        _ block: @escaping (_ context: CoreCryptoContextProtocol) async throws -> Result
    ) async throws -> Result
}

/// CoreCrypto client which manages one cryptographic client for proteus and MLS.
///
public final class CoreCrypto: CoreCryptoFfi, CoreCryptoProtocol, @unchecked Sendable {
    let database: Database?

    /// Initialize CoreCrypto with an Ffi instance
    private init(_ coreCrypto: WireCoreCryptoUniffi.CoreCryptoFfi, database: Database? = nil) {
        self.database = database
        // Due to Swift limitations, we can't use the `super` convenience intializer inside a child initializer, we
        // have to call the default `super` initializer. Luckily, we can do this meaningfully by cloning the pointer of
        // the instance that is passed in here.
        super.init(unsafeFromRawPointer: coreCrypto.uniffiClonePointer())
    }

    ///
    /// Initialise CoreCrypto with an encrypted key store.
    ///
    /// - Parameter keystorePath: path to the encrypted key store
    /// - Parameter key: secret key to unlock the encrypted key store
    ///
    public convenience init(database: Database) throws {
        let coreCrypto = try coreCryptoNew(database: database)
        self.init(coreCrypto, database: database)
    }

    @_documentation(visibility: private) required init(
        unsafeFromRawPointer pointer: UnsafeMutableRawPointer
    ) {
        self.database = nil
        super.init(unsafeFromRawPointer: pointer)
    }

    /// Instantiate a history client.
    ///
    /// This client exposes the full interface of `CoreCrypto`, but it should only be used to decrypt messages.
    /// Other use is a logic error.
    public static func historyClient(_ historySecret: HistorySecret) async throws -> CoreCrypto {
        let coreCrypto =
            try await WireCoreCryptoUniffi.coreCryptoHistoryClient(historySecret: historySecret)
        return self.init(coreCrypto)
    }

    public func transaction<Result>(
        _ block: @escaping (_ context: CoreCryptoContextProtocol) async throws -> Result
    ) async throws -> Result {
        let dbLocation = try await database?.getLocation()
        let filePath = dbLocation.map { FilePath(stringLiteral: $0) }
        let transactionExecutor = try TransactionExecutor<Result>(
            keystorePath: filePath, block)
        do {
            try await self.transaction(command: transactionExecutor)
        } catch {
            throw await transactionExecutor.innerError ?? error
        }
        return await transactionExecutor.result!
    }

    public func registerEpochObserver(_ epochObserver: EpochObserver) async throws {
        // we want to wrap the observer here to provide async indirection, so that no matter what
        // the observer that makes its way to the Rust side of things doesn't end up blocking
        try await self.registerEpochObserver(
            epochObserver: EpochObserverIndirector(epochObserver))
    }

    public func registerHistoryObserver(_ historyObserver: HistoryObserver) async throws {
        // we want to wrap the observer here to provide async indirection, so that no matter what
        // the observer that makes its way to the Rust side of things doesn't end up blocking
        try await self.registerHistoryObserver(
            historyObserver: HistoryObserverIndirector(historyObserver))
    }

    /// Returns the last resort PreKey id
    public static func proteusLastResortPrekeyId() throws -> UInt16 {
        return try WireCoreCryptoUniffi.proteusLastResortPrekeyIdFfi()
    }

    /// Proteus public key fingerprint
    /// It's basically the public key encoded as an hex string
    ///
    /// Returns Hex-encoded public key string
    public static func proteusFingerprintPrekeybundle(prekey: Data) throws -> String {
        return try WireCoreCryptoUniffi.proteusFingerprintPrekeybundleFfi(prekey: prekey)
    }
}

final class EpochObserverIndirector: EpochObserver {

    let epochObserver: EpochObserver

    init(_ epochObserver: EpochObserver) {
        self.epochObserver = epochObserver
    }

    func epochChanged(conversationId: ConversationId, epoch: UInt64) async throws {
        Task {
            try await epochObserver.epochChanged(conversationId: conversationId, epoch: epoch)
        }
    }

}

final class HistoryObserverIndirector: HistoryObserver {

    let historyObserver: HistoryObserver

    init(_ historyObserver: HistoryObserver) {
        self.historyObserver = historyObserver
    }

    func historyClientCreated(conversationId: ConversationId, secret: HistorySecret) async throws {
        Task {
            try await historyObserver.historyClientCreated(
                conversationId: conversationId, secret: secret)
        }
    }
}

final actor TransactionExecutor<Result>: WireCoreCryptoUniffi.CoreCryptoCommand {

    let block: (_ context: CoreCryptoContextProtocol) async throws -> Result
    var result: Result?
    var innerError: Error?
    var fileDescriptor: FileDescriptor?

    init(
        keystorePath: FilePath?,
        _ block: @escaping (_ context: CoreCryptoContextProtocol) async throws -> Result
    ) throws {
        self.block = block

        if let keystorePath {
            let path = keystorePath.absolutePath().removingLastComponent()
            self.fileDescriptor = try FileDescriptor.open(path, .readOnly, options: .directory)
        }
    }

    deinit {
        try? fileDescriptor?.close()
    }

    func execute(context: WireCoreCryptoUniffi.CoreCryptoContext) async throws {
        // Only aquire lock if we are using an instance which persists to disk
        if fileDescriptor != nil {
            acquireFileLock()

            // Reload any cached proteus sessions from disk since they may be invalid
            try await context.proteusReloadSessions()
        }

        defer {
            releaseFileLock()
        }

        do {
            result = try await block(context)
        } catch {
            innerError = error
            throw error
        }
    }

    func acquireFileLock() {
        guard let fileDescriptor else { return }
        let result = flock(fileDescriptor.rawValue, LOCK_EX)
        if result != 0 {
            fatalError("Failed aquire lock: \(errno))")
        }
    }

    func releaseFileLock() {
        guard let fileDescriptor else { return }
        let result = flock(fileDescriptor.rawValue, LOCK_UN)
        if result != 0 {
            fatalError("Failed release lock: \(errno))")
        }
    }

}

extension Credential {
    /// Construct a new Credential from ciphersuite and client id
    public static func basic(
        ciphersuite: Ciphersuite,
        clientId: ClientId
    ) throws -> Credential {
        return try credentialBasic(ciphersuite: ciphersuite, clientId: clientId)
    }
}

extension Database {
    /// Initialise or open a Database.
    ///
    /// - Parameter location: path to the database
    /// - Parameter key: secret key to unlock the database
    ///
    public static func open(location: String, key: DatabaseKey) async throws -> Database {
        return try await openDatabase(location: location, key: key)
    }

    /// Initialise an in-memory Database whose data will be lost when the instance is dropped.
    ///
    /// - Parameter key: secret key to unlock the database
    public static func open(key: DatabaseKey) async throws -> Database {
        return try await inMemoryDatabase(key: key)
    }
}

File: crypto-ffi/bindings/shared/src/commonMain/kotlin/com/wire/crypto/CoreCrypto.kt
@file:Suppress("TooGenericExceptionCaught")

package com.wire.crypto

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.ByteArray

/** Wrap a `CoreCrypto` instance in a `CoreCryptoClient` instance. Should largely be invisible to end-users. */
fun CoreCryptoFfi.lift() = CoreCrypto(this)

/**
 * A high-level wrapper around a CoreCrypto client as emitted by Uniffi.
 *
 * This wrapper should be largely transparent to end users. It exists to improve the
 * callback interfaces: `.transaction(...)`, `.registerFooObserver(...)`, etc.
 */
class CoreCrypto(private val cc: CoreCryptoFfi) {
    companion object {
        /** Opens a core crypto client with the specified database, previously instantiated via [openDatabase].  */
        suspend operator fun invoke(
            database: Database,
        ) =
            CoreCrypto(coreCryptoNew(database))

        /**
         * Instantiate a history client.
         *
         * This client exposes the full interface of `CoreCrypto`, but it should only be used to decrypt messages.
         * Other use is a logic error.
         */
        suspend fun historyClient(historySecret: HistorySecret) = CoreCrypto(coreCryptoHistoryClient(historySecret))

        /**
         * @returns The last resort PreKey id
         */
        @Throws(CoreCryptoException::class)
        fun proteusLastResortPrekeyId() = proteusLastResortPrekeyIdFfi()

        /**
         * Hex-encoded fingerprint of the given prekey
         *
         * @param prekey - the prekey bundle to get the fingerprint from
         * @returns Hex-encoded public key string
         **/
        @Throws(CoreCryptoException::class)
        fun proteusFingerprintPrekeybundle(
            prekey: kotlin.ByteArray
        ) = proteusFingerprintPrekeybundleFfi(prekey)
    }

    /**
     * Starts a [NonCancellable] transaction in Core Crypto. If the callback succeeds, it will be committed,
     * otherwise, every operation performed with the context will be discarded.
     *
     * Check [Uniffi's documentation](https://mozilla.github.io/uniffi-rs/latest/futures.html#cancelling-async-code)
     * about async code, that mentions that it does not support cancellation. So we go around it by not cancelling
     * it either.
     * @param R the type returned by the transaction block
     * @param block the function to be executed within the transaction context.
     *              A [CoreCryptoContext] will be given as parameter to this function.
     *
     * @return the return of the function passed as parameter
     */
    @Suppress("unchecked_cast")
    suspend fun <R> transaction(block: suspend (context: CoreCryptoContext) -> R): R = withContext(NonCancellable) {
        var result: R? = null
        var error: Throwable? = null
        try {
            this@CoreCrypto.cc.transaction(object : CoreCryptoCommand {
                override suspend fun execute(context: CoreCryptoContext) {
                    try {
                        result = block(context)
                    } catch (e: Throwable) {
                        // We want to catch the error before it gets wrapped by core crypto.
                        error = e
                        // This is to tell core crypto that there was an error inside the transaction.
                        throw e
                    }
                }
            })
        } catch (e: Throwable) {
            // We prefer the closure error if it's available since the transaction won't include it
            error = error ?: e
        }
        if (error != null) {
            throw error as Throwable
        }

        // Since we know that the transaction will either succeed or throw it's safe to do an unchecked cast here
        return@withContext result as R
    }

    /**
     * Register an Epoch Observer which will be notified every time a conversation's epoch changes.
     *
     * This function should be called 0 or 1 times in the lifetime of CoreCrypto, regardless of the number of transactions.
     */
    suspend fun registerEpochObserver(scope: CoroutineScope, epochObserver: EpochObserver) {
        // we want to wrap the observer here to provide async indirection, so that no matter what
        // the observer that makes its way to the Rust side of things doesn't end up blocking
        val observerIndirector = object : EpochObserver {
            override suspend fun epochChanged(conversationId: ConversationId, epoch: kotlin.ULong) {
                scope.launch { epochObserver.epochChanged(conversationId, epoch) }
            }
        }
        return cc.registerEpochObserver(observerIndirector)
    }

    /**
     * Register a History Observer which will be notified every time a new history client is created.
     *
     * This function should be called 0 or 1 times in the lifetime of CoreCrypto, regardless of the number of transactions.
     */
    suspend fun registerHistoryObserver(scope: CoroutineScope, historyObserver: HistoryObserver) {
        // we want to wrap the observer here to provide async indirection, so that no matter what
        // the observer that makes its way to the Rust side of things doesn't end up blocking
        val observerIndirector = object : HistoryObserver {
            override suspend fun historyClientCreated(
                conversationId: ConversationId,
                secret: HistorySecret
            ) {
                scope.launch { historyObserver.historyClientCreated(conversationId, secret) }
            }
        }
        return cc.registerHistoryObserver(observerIndirector)
    }

    /**
     * See [CoreCryptoContext.isHistorySharingEnabled]
     *
     * @param conversationId conversation identifier
     * @return true if history sharing is enabled
     */
    suspend fun isHistorySharingEnabled(conversationId: ConversationId): Boolean = cc.isHistorySharingEnabled(conversationId)

    /**
     * Closes this [CoreCrypto] instance and deallocates all loaded resources.
     *
     * **CAUTION**: This {@link CoreCrypto} instance won't be usable after a call to this method, but there's no way to express this requirement in Kotlin, so you'll get errors instead!
     */
    fun close() {
        cc.close()
    }
}
